### PR TITLE
feat: add garbagecollector

### PR DESCRIPTION
**Pull request by cndoit18**
fixes: #588

### CODE REVIEW

File: pkg/graph/graph.go [hunk 1]
```diff
@@ -46,21 +81,78 @@ func (rm *resourceManager) AddResource(ctx context.Context, resource Resource, b
...
+			},
+			GroupVersionKind: gvk,
+		}
+
+		if _, ok := rm.primary[key]; ok {
+			return false, nil
+		}
+
+		rm.primary[key] = res
+	case inlineObject:
```
Comment (heww) [inline pkg/graph/graph.go:131]: This interface only appears here once, we don't need to create an interface type, just use `case interface{GetObject() client.Object}`

Comment (cndoit18) [reply pkg/graph/graph.go]: thx, updated

File: pkg/controller/common.go [hunk 2]
```diff
@@ -64,6 +70,48 @@ func NewController(ctx context.Context, base controllers.Controller, rm Resource
...
+					},
+				},
+			}
+			if err := c.Client.Create(ctx, sar); err != nil {
+				return err
+			}
+
+			if sar.Status.Allowed {
+				c.deletableGVKs[gv.WithKind(rl.APIResources[i].Kind)] = struct{}{}
+			}
```
Comment (bitsf) [inline pkg/controller/common.go:112]: how about logging exceptions with `else`

Comment (cndoit18) [reply pkg/controller/common.go]: should not be logging exceptions, the effect here is similar to filter

**Review by heww — APPROVED** [review]

**Review by bitsf — APPROVED** [review]

File: pkg/graph/graph.go [hunk 3]
```diff
@@ -22,6 +22,19 @@ func NewResourceManager() Manager {
 	}
 }
 
+func (rm *resourceManager) GetAllResource(ctx context.Context) []Resource {
```
Comment (chlins) [inline pkg/graph/graph.go:25]: GetAllResources?

File: pkg/graph/graph.go [hunk 4]
```diff
@@ -22,6 +22,19 @@ func NewResourceManager() Manager {
...
+	resources := []Resource{}
+
+	rm.lock.RLock()
+	defer rm.lock.RUnlock()
+
+	for resource := range rm.resources {
+		resources = append(resources, resource)
```
Comment (chlins) [inline pkg/graph/graph.go:32]: why we need to copy them from `rm.resources` to a new slice other than use or return `rm.resources` directly?

**Review by chlins — APPROVED** [review]
/lgtm

File: pkg/graph/graph.go (hunk 4 above)
Comment (cndoit18) [reply pkg/graph/graph.go:32]: Because the structure of this field is `map[Resource][]Resource`

File: pkg/graph/graph.go (hunk 3 above)
Comment (cndoit18) [reply pkg/graph/graph.go]: Thx, updated

File: pkg/graph/graph.go (hunk 4 above)
Comment (chlins) [reply pkg/graph/graph.go:32]: Thx, got it.

**Review by steven-zou — APPROVED** [review]
LGTM

**Review by yxxhero — APPROVED** [review]

**Pull request merged**